espresso/environment/4_confirmation_integrity_with_reorgs_test.go

```diff
@@ -0,0 +1,217 @@
+		batches = append(batches, batchHash)
+
+	}
+	return batches
+}
+
+// Main logic of the test:
+// 1. Collect some unsafe batches in list L.
+// 2. Do the reorg.
+// 3. Collect the batches into list L'.
+// 4. Check that L=L'.
+func run(ctx context.Context, t *testing.T, system *e2esys.System) {
+	l2Seq := system.NodeClient(e2esys.RoleSeq)
+	l2Verif := system.NodeClient(e2esys.RoleVerif)
+	l1Client := system.NodeClient(e2esys.RoleL1)
+
+	var unsafeL2Height uint64
+	var l1Height uint64
+
+	// Wait for batcher to start advancing L2 head
+	_, err := geth.WaitForBlockToBeSafe(big.NewInt(2), l2Seq, 2*time.Minute)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("L2 isn't progressing:\nhave:\n\t%v\nwant:\n\t%v", have, want)
+	}
+
+	t.Log("L2 is progressing")
+
+	// Fetch batches before reorg
+	batchesBefore, L1BlockHeightToReorgTo, startIndex := collectBatchesPublishedOnUnfinalizedL1Blocks(ctx, t, system, l1Client, l2Seq, l2Verif)
+
+	l1Origin, err := l1Client.BlockByNumber(ctx, new(big.Int).SetUint64(L1BlockHeightToReorgTo))
+	require.NoError(t, err)
+
+	log.Info("+++ L2 blocks before reorg", "value", batchesBefore)
+	// Introduce a reorg at L1
+	l1Height, err = l1Client.BlockNumber(ctx)
+	require.NoError(t, err)
+	t.Logf("Introducing reorg at L1Origin %d, L1Head %d, l2Head %d", l1Origin.Number(), l1Height, unsafeL2Height)
+	err = system.ForkL1(l1Origin.ParentHash())
+	require.NoError(t, err)
+
+	n := len(batchesBefore)
+	batchesAfter := collectFirstNL2SafeBlocks(ctx, t, system, l2Verif, n, startIndex)
+
+	log.Info("+++ L2 blocks after reorg", "value", batchesAfter)
+
+	assert.Equal(t, batchesAfter, batchesBefore)
+
+}
+
+// TestConfirmationIntegrityWithReorgs
+// Post batches to both Espresso and the L1 then force the L1 to reorg back to an earlier state in which those batches have not been posted.
+// Wait for some time and check that the batches are eventually posted to the L1 again, in the same order as they were originally sequenced,
+// as if the reorg did not happen.
+// More specifically the test is defined as follows
+//
+//	Arrange:
+//		Running Sequencer, Batcher in Espresso mode, OP node.
+//	Act:
+//		Store the (unfinalized) head of the L1 in variable h.
+//		Wait for the first n batches to be posted on possibly unfinalized L1 blocks.
+//		Collect the batches of the corresponding batches and store them in list L.
+//		Reorg the L1 to height h.
+//		Wait for the L2 to reach safe height n again as the corresponding batches are submitted again to L1.
+//		Store these n batches in L'
+//	Assert:
+//		L == L'
+func TestConfirmationIntegrityWithReorgs(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	launcher := new(env.EspressoDevNodeLauncherDocker)
+
+	system, _, err := launcher.StartDevNet(ctx, t, 12)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start dev environment with espresso dev node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
```

Comment:
```suggestion
	require.NoError(t, err, "failed to start dev environment with espresso dev node")
```
Similar suggestion as above.